flip the env to the correct side when intersect 2 typevar

N5N3 opened the pull request:
The intersection result should be order-independent. But our subtype algorithm is sensitive to which side the typevar comes from.
This PR tries to update the `right`s in env based on the following subtype check to avoid some false `Union{}`.

fix #41561.

N5N3 committed:
typeintersect: flip the env to the "correct" side when intersect 2 typevar

N5N3 commented:
@nanosoldier `runtests()`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/48006#issuecomment-1366340228) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/b64624b_vs_a9e0545/report.html).

N5N3 committed:
support nested `fresh_subtype_in_env`

N5N3 committed:
spread the usage of `fresh_subtype_in_env`

this helps to get more precise intersect result.

N5N3 commented:
@nanosoldier runtests()

N5N3 commented:
@nanosoldier `runtests()`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/48006#issuecomment-1368223839) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/2242ecc_vs_a9e0545/report.html).

N5N3 committed:
more accurate `merge_env`

This fix:
1. don't ignore occurs_inv/cov (they should be merged by max.)
2. only merge typevars in the current union decision.
3. add a check before `env` merge to avoid circulation.

N5N3 committed:
Set `ub` more carefully in `intersect_var`

At present, `intersect_var` uses `intersect_aside` to update the var's upbound.
Since the inner `intersect` can't access the var, the returned bound might contain circulation.
This commit tries to remove the invalid part and keep the env clean.

N5N3 commented:
@nanosoldier `runtests(["PlantSimEngine", "RRTMGP", "DynamicGrids", "TensorBoardLogger", "Causal", "PlantMeteo", "FastaLoader", "LicenseGrabber", "OhMyArtifacts"])`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/48006#issuecomment-1369118729) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/c5aff3a_vs_7a561bd/report.html).

N5N3 committed:
Add another fastpath to avoid stack-overflow.

N5N3 committed:
Avoid stack-overflow in `reachable_var`

N5N3 commented:
@nanosoldier `runtests()`

N5N3 committed:
Fix the reported error when we meet `1<:x<:Any`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/48006#issuecomment-1371500349) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/2c7c9d9_vs_4a42367/report.html).
